### DISCUSSION TITLE
Alias for Calling AET

### DISCUSSION THREAD

**Discussion by fonteneu** [dcm4che/dcm4chee-arc-light#4596] [General]
I am trying to migrate from dcm4chee 2.x to dcm4chee arc light. However I'm blocked by the following (afaict) missing feature.

I need to accept MWL and MPPS requests from modality with AET "MOD1" but dcm4chee only knows about worklist for "MOD2". This is important because facilities occasionally want extra modalities to access the same worklist.

In dcm4chee 2.x this was made possible by providing a `mwl-cfindrq.xsl` in a directory whose name is "MOD1", and the XSL file then hardcodes the calling AET to the real one, "MOD2".

I noticed https://github.com/dcm4che/dcm4chee-arc-light/issues/1442 but it does not seem to cover this use case.

**Comment by vrindanayak** [comment]
> I need to accept MWL and MPPS requests from modality with AET "MOD1" but dcm4chee only knows about worklist for "MOD2". This is important because facilities occasionally want extra modalities to access the same worklist.

How was the worklist created in DCM4CHEE? The archive supports accepting multiple values in ORC.18 of (ORM^O01 or OMG^O19) or IPC.9 of (OMI^O23). If the worklist already has multiple Scheduled Station AE Titles, the calling AET of modality can be coerced in MWL C-FIND request using a coercion.

**Reply by fonteneu** [reply]
> If the worklist already has multiple Scheduled Station AE Titles, the calling AET of modality can be coerced in MWL C-FIND request using a coercion.

The worklist does not and will not contain multiple scheduled AE titles. That's not under our control, and thus not a viable solution. 

That's why we rely on `mwl-cfindrq.xsl`

Example: high study load or equipment issues causes a different modality to be temporarily assigned to pull the same worklist as another modality. Sometimes multiple modalities pull worklists for another AET. This situation can last for just a few hours or months.

When the need arise, we add a mwl-cfindrq.xsl to create an alias such that both MOD1 and MOD2 will get MOD2s worklist. 

Specifically the file `MOD1/mwl-cfindrq.xsl` sets the SPSS / Scheduled AET to `MOD2`

This works very well, but cannot find similar in arc light.

**Comment by vrindanayak** [comment]
You can achieve this using a 2-step coercion

### LDAP Config

```
version: 1

dn: cn=coerce-ss-aet-mwl-4596,dicomDeviceName=dcm4chee-arc,cn=Devices,cn=DICOM Configuration,dc=dcm4che,dc=org
objectClass: dcmArchiveAttributeCoercion2
cn: coerce-ss-aet-mwl-4596
dcmDIMSE: C_FIND_RQ
dcmURI: use-calling-aet-as:ScheduledStationAETitle
dicomTransferRole: SCU
dcmProperty: SendingApplicationEntityTitle=MOD1
dcmRulePriority: 1
dcmSOPClass: 1.2.840.10008.5.1.4.31

dn: cn=coerce-ss-aet-mwl2-4596,dicomDeviceName=dcm4chee-arc,cn=Devices,cn=DICOM Configuration,dc=dcm4che,dc=org
objectClass: dcmArchiveAttributeCoercion2
cn: coerce-ss-aet-mwl2-4596
dcmDIMSE: C_FIND_RQ
dcmURI: merge-attrs:
dicomTransferRole: SCU
dcmRulePriority: 0
dcmMergeAttribute: ScheduledProcedureStepSequence.ScheduledStationAETitle=MOD2
dcmProperty: SendingApplicationEntityTitle=MOD1
dcmSOPClass: 1.2.840.10008.5.1.4.31
```

If you have more modalities' AETs which need the same functionality, just pipe the values in the `dcmProperty`, eg. 
```
dcmProperty: SendingApplicationEntityTitle=MOD1|MOD3|MOD4
```

### Verification

```
findscu -M MWL -b MOD1 -c WORKLIST@localhost:11112 -r ScheduledProcedureStepSequence -r AccessionNumber
11:57:10.906 INFO  - Initiate connection from 0.0.0.0/0.0.0.0:0 to localhost:11112
..........
..........
11:57:10.975 DEBUG - MOD1->WORKLIST(1) >> 1:C-FIND-RSP Dataset:
(0008,0005) CS [] SpecificCharacterSet
(0008,0050) SH [ACC1231] AccessionNumber
(0040,0100) SQ [1 Items] ScheduledProcedureStepSequence
>Item #1
>(0008,0060) CS [CR] Modality
>(0040,0001) AE [MOD2] ScheduledStationAETitle
>(0040,0002) DA [20000816] ScheduledProcedureStepStartDate
>(0040,0003) TM [151515] ScheduledProcedureStepStartTime
>(0040,0006) PN [] ScheduledPerformingPhysicianName
>(0040,0007) LO [Protocol~Meaning] ScheduledProcedureStepDescription
>(0040,0008) SQ [1 Items] ScheduledProtocolCodeSequence
>>Item #1
>>(0008,0100) SH [ProtocolCode] CodeValue
>>(0008,0102) SH [ProtocolScheme] CodingSchemeDesignator
>>(0008,0104) LO [Protocol~Meaning] CodeMeaning
>(0040,0009) SH [RPID] ScheduledProcedureStepID
>(0040,0010) SH [MOD2_Name] ScheduledStationName
>(0040,0011) SH [SPSL] ScheduledProcedureStepLocation
>(0040,0020) CS [SCHEDULED] ScheduledProcedureStepStatus
```

In archive server log
```
11:57:10,916 INFO  [org.dcm4che3.net.Connection] (EE-ManagedExecutorService-default-Thread-2) Accept connection Socket[addr=/127.0.0.1,port=58319,localport=11112]
11:57:10,929 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15) WORKLIST<-MOD1(4) >> A-ASSOCIATE-RQ
11:57:10,929 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15) WORKLIST<-MOD1(4) << A-ASSOCIATE-AC
11:57:10,956 INFO  [org.dcm4che3.net.Dimse] (EE-ManagedExecutorService-default-Thread-15) WORKLIST<-MOD1(4) >> 1:C-FIND-RQ[pcid=1, prior=0
  cuid=1.2.840.10008.5.1.4.31 - Modality Worklist Information Model - FIND
  tsuid=1.2.840.10008.1.2 - Implicit VR Little Endian]
11:57:10,957 INFO  [org.dcm4chee.arc.procedure.scp.MWLCFindSCP] (EE-ManagedExecutorService-default-Thread-15) WORKLIST<-MOD1(4): Process MWL C-FIND RQ:
(0008,0050) SH [] AccessionNumber
(0040,0100) SQ [1 Items] ScheduledProcedureStepSequence
...

11:57:10,957 INFO  [org.dcm4chee.arc.coerce.impl.UseCallingAETAsCoercionProcessor] (EE-ManagedExecutorService-default-Thread-15) Filter MWL by Calling AET by ArchiveAttributeCoercion2[cn=coerce-ss-aet-mwl-4596, priority=1, DIMSE=C_FIND_RQ, role=SCU, cuids=[1.2.840.10008.5.1.4.31], conditions={SendingApplicationEntityTitle=MOD1}, uri=use-calling-aet-as:ScheduledStationAETitle, description=null, onFailure=RETHROW, sufficient=false, attributeUpdatePolicy=MERGE, mergeAttributes=[], deviceCoercionParam=null, otherCoercionParams={}]
11:57:10,957 INFO  [org.dcm4chee.arc.coerce.impl.MergeAttributesCoercionProcessor] (EE-ManagedExecutorService-default-Thread-15) Merge Attributes using ArchiveAttributeCoercion2[cn=coerce-ss-aet-mwl2-4596, priority=0, DIMSE=C_FIND_RQ, role=SCU, cuids=[1.2.840.10008.5.1.4.31], conditions={SendingApplicationEntityTitle=MOD1}, uri=merge-attrs:, description=null, onFailure=RETHROW, sufficient=false, attributeUpdatePolicy=MERGE, mergeAttributes=[ScheduledProcedureStepSequence.ScheduledStationAETitle=MOD2], deviceCoercionParam=null, otherCoercionParams={}]
11:57:10,964 INFO  [org.dcm4che3.net.Dimse] (EE-ManagedExecutorService-default-Thread-12) WORKLIST<-MOD1(4) << 1:C-FIND-RSP[pcid=1, status=ff00H
  cuid=1.2.840.10008.5.1.4.31 - Modality Worklist Information Model - FIND
  tsuid=1.2.840.10008.1.2 - Implicit VR Little Endian]
11:57:10,965 INFO  [org.dcm4che3.net.Dimse] (EE-ManagedExecutorService-default-Thread-12) WORKLIST<-MOD1(4) << 1:C-FIND-RSP[pcid=1, status=0H
  cuid=1.2.840.10008.5.1.4.31 - Modality Worklist Information Model - FIND
  tsuid=1.2.840.10008.1.2 - Implicit VR Little Endian]
11:57:10,976 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15) WORKLIST<-MOD1(4) >> A-RELEASE-RQ
11:57:10,976 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15) WORKLIST<-MOD1(4) << A-RELEASE-RP
11:57:11,027 INFO  [org.dcm4che3.net.Association] (EE-ManagedScheduledExecutorService-default-Thread-6) WORKLIST<-MOD1(4): close Socket[addr=view-localhost/127.0.0.1,port=58319,localport=11112]
```

**Comment by fonteneu** [comment]
@vrindanayak thank you!

Is this LDAP config documented anywhere?

**Reply by fonteneu** [reply]
For instance, is coerce-ss-aet-mwl-4596 an arbitrary id or does it have any significance?

**Reply by vrindanayak** [reply]
> @vrindanayak thank you!
> 
> Is this LDAP config documented anywhere?

No - it is specific to your use case.

> For instance, is coerce-ss-aet-mwl-4596 an arbitrary id or does it have any significance?

`coerce-ss-aet-mwl-4596` is an arbitrary name I gave to the coercion rule for testing. You may rename it as you like.

**Reply by fonteneu** [reply]
Sounds good, your information is very helpful.

I have a final followup about my transition to dcm4chee arc if you don't mind:

1. Do I set CALLED AET in LDAP config as well? This is the DICOM server's AET. We used JMX for this in dcm4chee 2.x. Do you have an example?

2. Is the JMX config UI gone altogether and replaced with LDAP?

3. Is there a way to automatically delete old worklist and mpps entries? Say, older than a week.

**Reply by vrindanayak** [reply]
> 1. Do I set CALLED AET in LDAP config as well? This is the DICOM server's AET. We used JMX for this in dcm4chee 2.x. Do you have an example?

I do not understand what you are implying here. Dcm4chee archive 5.x already has AEs configured by default. As you can see in the test verification example above, modality worklists can be invoked using the `WORKLIST` AE of archive

> 2. Is the JMX config UI gone altogether and replaced with LDAP?

I'm not aware of JMX, never worked on dcm4chee archive 2.x.

> 3. Is there a way to automatically delete old worklist and mpps entries? Say, older than a week.

There is a feature to automatically delete MWLs, but not for MPPS. For auto deletion of MWLs, you may configure [Delete MWL Delay(s)](https://dcm4chee-arc-cs.readthedocs.io/en/latest/networking/config/archiveDevice.html#dcmdeletemwldelay) and [MWL Polling Interval](https://dcm4chee-arc-cs.readthedocs.io/en/latest/networking/config/archiveDevice.html#dcmmwlpollinginterval)

**Reply by fonteneu** [reply]
> I do not understand what you are implying here.

I need to set the server's AET to something specific. 

2.x called this "CalledAETitles - accepted called AE titles" in the modality worklist SCP configuration

**Reply by vrindanayak** [reply]
> I need to set the server's AET to something specific.
> 
> 2.x called this "CalledAETitles - accepted called AE titles" in the modality worklist SCP configuration

There's a config for [Accepted Calling AE Title(s)](https://dcm4chee-arc-cs.readthedocs.io/en/latest/networking/config/dcmNetworkAE.html#dcmacceptedcallingaetitle) but not vice versa, i.e. which clients (modalities) will be accepted.

The dcm4chee archive (server's) AE (in this case `WORKLIST`) - shouldn't this be configured in the modality?

**Reply by fonteneu** [reply]
Unfortunately the facility we install at has requirements for what the server AE must be. 

They already have tens of modalities configured with it and we cannot change it.

If called (server) AET is hardcoded to WORKLIST then that's a blocker for sure. Similarly for MPPS - dcm4chee 2.x let us set called AET to whatever we chose.

**Reply by fonteneu** [reply]
Seems possible: https://groups.google.com/g/dcm4che/c/alP8PThcJo8

**Reply by vrindanayak** [reply]
> Seems possible: https://groups.google.com/g/dcm4che/c/alP8PThcJo8

Yeah, this post explains how you can change the default AEs of archive. It was not clear earlier that you want to change the AE of the archive - see [AE Title(s) of the Archive](https://github.com/dcm4che/dcm4chee-arc-light/wiki/AE-Title%28s%29-of-the-Archive)

**Comment by fonteneu** [comment]
I hope dcm4chee can add a simpler way to accomplish this. This is proving difficult to get right for clients.

**Comment by vrindanayak** [comment]
@fonteneu : You had a question (arrived on my email) - I no longer see it in this Discussion
> [@vrindanayak](https://github.com/vrindanayak) can I use a tool like phpLDAPadmin to make these changes? What tool do you recommend?
> LDAP is new to me and this looks complicated for what was previously a simple xml change

You can easily import ldifs in to your configuration using [Apache Directory Studio](https://directory.apache.org/studio/) client or you may use `ldapadd` / `ldapmodify` using command line. For reference, see https://github.com/dcm4che/dcm4chee-arc-light/wiki/Installation#openldap-with-dynamic-runtime-configuration
